Lib/inspect.py

```diff
@@ -1898,7 +2015,7 @@ def _signature_is_functionlike(obj):
             isinstance(name, str) and
             (defaults is None or isinstance(defaults, tuple)) and
             (kwdefaults is None or isinstance(kwdefaults, dict)) and
-            isinstance(annotations, dict))
+            isinstance(annotations, (dict, None)))
```

Comment:
`isinstance(obj, None)` doesn't work.

Reply:
Interestingly it gets checked lazily, which is probably why this didn't get picked up by tests:
```
>>> isinstance(1, (dict,None))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: isinstance() arg 2 must be a type, a tuple of types or a union
>>> isinstance({}, (dict,None))
True
>>> 
```
I'm not sure when a function's `__annotations__` would be None.

Reply:
Oh!  That `isinstance` `None` stuff sure is dumb.  Thanks for spotting it!

Currently I don't know why someone would set `__annotations__`  to `None` either, but I note that it is specifically permitted by the function object.  It only allows setting `__annotations__` to a dict or `None`.  (Classes and modules are agnostic and will permit setting `__annotations__` to any value.)

It's probably too late for Python 3, but if there ever is a Python 4, I think `__annotations__` should enforce always being either a dict or `None` on all objects that support it, and that it should default to `None` when the object has no annotations.

Reply:
Also note just a few lines above:

    annotations = getattr(obj, '__annotations__', None)

If `__annotations__` isn't set, `annotations` will be `None`.  It's true that `__annotations__` is always set on CPython function objects, but that isn't guaranteed of other callables / objects that may be tested by this function.

Reply:
Oh, it's more complicated than I remembered.  You're allowed to set `func.__annotations__ = None`, but it internally just sets its annotations storage to `NULL`, and then if you retrieve annotations it lazy-creates a new dict.  So `func.__annotations__` on a real CPython function object will never ever be anything but a dict.  (Again, not so for classes, modules, wrapped functions, etc ad nauseum.)

And I just fixed this bug and refreshed the PR.  Marking this as "resolved".